File: Porting/Source/R_API_piragnolo/getSpatialData_dev.R
library(httr)
library(xml2)
source("getSpatialData_dev_internal.R")



cophub_api<-function (x, p, user, pw) 
{
  if (x == "auto") {
    if (p == "Sentinel-1" | p == "Sentinel-2") {
      x <- "operational"
    }
    else {
      x <- "pre-ops"
    }
  }
  if (x == "operational") {
    x <- getOption("gSD.api")$dhus
  }
  if (x == "pre-ops") {
    x <- getOption("gSD.api")$s3
    user <- "s3guest"
    pw <- "s3guest"
  }
  return(c(user, pw, x))
}




set_archive<-function (dir_archive) 
{
  if (!is.character(dir_archive)) {
    out(paste0("Argument 'dir_archive' needs to be of type 'character'."), 
        type = 3)
  }
  if (!dir.exists(dir_archive)) 
    out("The defined directory does not exist.", type = 3)
  options(gSD.archive = dir_archive)
  options(gSD.archive_set = TRUE)
}



login_CopHub<-function (username, password = NULL) 
{
  if (is.null(password)) {
    password <- getPass()
  }
  char_args <- list(username = username, password = password)
  for (i in 1:length(char_args)) {
    if (!is.character(char_args[[i]])) {
      out(paste0("Argument '", names(char_args[i]), "' needs to be of type 'character'."), 
          type = 3)
    }
  }
  options(gSD.cophub_user = username)
  options(gSD.cophub_pass = password)
  options(gSD.cophub_set = TRUE)
}


getSentinel_query2<-function (time_range, platform, aoi = NULL, username = NULL, 
          password = NULL, hub = "auto", verbose = TRUE, ingestiondate=FALSE, extra=NULL) 
{
  if (is.TRUE(getOption("gSD.cophub_set"))) {
    if (is.null(username)) 
      username <- getOption("gSD.cophub_user")
    if (is.null(password)) 
      password <- getOption("gSD.cophub_pass")
  }
  if (!is.character(username)) 
    out("Argument 'username' needs to be of type 'character'. You can use 'login_CopHub()' to define your login credentials globally.", 
        type = 3)
  if (!is.null(password)) {
    password = password
  }
  else {
    password = getPass()
  }
  if (inherits(verbose, "logical")) 
    options(gSD.verbose = verbose)
  # if (is.null(aoi)) {
  #   if (is.TRUE(getOption("gSD.aoi_set"))) {
  #     aoi <- getOption("gSD.aoi")
  #   }
  #   else {
  #     out("Argument 'aoi' is undefined and no session AOI could be obtained. Define aoi or use set_aoi() to define a session AOI.", 
  #         type = 3)
  #   }
  # }
  #aoi <- make_aoi(aoi, type = "matrix")
  char_args <- list(time_range = time_range, platform = platform)
  for (i in 1:length(char_args)) if (!is.character(char_args[[i]])) 
    out(paste0("Argument '", names(char_args[i]), "' needs to be of type 'character'."), 
        type = 3)
  if (length(time_range) != 2) {
    out("Argument 'time_range' must contain two elements (start and stop time).", 
        type = 3)
  }
  cop.url <- function(ext.xy, url.root, platform, time.range, 
                      row.start) {
   if(!ingestiondate) qs <- list(url.root = paste0(url.root, "/"), search = c("search?start=", 
                                                            "&rows=100&q=("), 
               and = "%20AND%20", 
               aoi.poly = c("footprint:%22Intersects(POLYGON((", ")))%22"), 
               platformname = "platformname:", 
               time = list(`[` = "beginposition:%5b", to = "%20TO%20", `]` = "%5d"))
   else  qs <- list(url.root = paste0(url.root, "/"), search = c("search?start=", 
                                                                 "&rows=100&q=("), 
                    and = "%20AND%20", 
                    aoi.poly = c("footprint:%22Intersects(POLYGON((", ")))%22"), 
                    platformname = "platformname:", 
                    time = list(`[` = "ingestiondate:%5b", to = "%20TO%20", `]` = "%5d"))
   # aoi.str <- paste0(apply(ext.xy, MARGIN = 1, function(x) paste0(x, 
   #                                                                collapse = "%20")), collapse = ",")
    time.range <- sapply(time.range, function(x) {
      tt<-as.POSIXct(time.range)
     # paste0(x, "T00:00:00.000Z")
      format(tt, "%Y-%m-%dT%H:%M:%S.000Z" )
      }, USE.NAMES = F)
 
     
  if(is.null(extra)) urlstring<-paste0(qs$url.root, qs$search[1], toString(row.start),
                      qs$search[2], qs$platformname, platform,  qs$and, qs$time$`[`, 
                      time.range[1], qs$time$to, time.range[2], qs$time$`]`, 
                      ")")
     
  else  urlstring<-paste0(qs$url.root, qs$search[1], toString(row.start),
                      qs$search[2], qs$platformname, platform,  qs$and, extra,  qs$and, qs$time$`[`, 
                      time.range[1], qs$time$to, time.range[2], qs$time$`]`, 
                      ")")
    print(urlstring)
    print(as.character(Sys.time()))
    return(urlstring)
  } 
  cred <- cophub_api(hub, platform, username, password)
  row.start <- -100
  re.query <- T
  give.return <- T
  query.list <- list()
  while (is.TRUE(re.query)) {
    row.start <- row.start + 100
    query <- gSD.get(cop.url(ext.xy = aoi, url.root = cred[3], 
                             platform = platform, time.range = time_range, row.start = row.start), 
                     cred[1], cred[2])
    query.xml <- suppressMessages(xml_contents(content(query)))
    query.list <- c(query.list, lapply(query.xml[grep("entry", 
                                                      query.xml)], function(x) xml_contents(x)))
    if (length(query.list) == 0 & row.start == 0) {
      out("No results could be obtained for this request.", 
          msg = T)
      re.query <- F
      give.return <- F
    }
    if (length(query.list) != row.start + 100) 
      re.query <- F
  }
  if (is.TRUE(give.return)) {
    field.tag <- c("title", "link href", "link rel=\"alternative\"", 
                   "link rel=\"icon\"", "summary", "name")
    field.names <- c("title", "url", "url.alt", "url.icon", 
                     "summary")
    query.cont <- lapply(query.list, function(x, field = field.tag) unlist(lapply(field, 
                                                                                  function(f, y = x) grep(f, y, value = T))))
    query.names <- lapply(query.cont, function(x, field.n = field.names) c(field.n, 
                                                                           sapply(x[(length(field.n) + 1):length(x)], function(y) strsplit(y, 
                                                                                                                                           "\"")[[1]][2], USE.NAMES = F)))
    query.fields <- lapply(query.cont, function(x) sapply(x, 
                                                          function(y) strsplit(strsplit(y, ">")[[1]][2], "<")[[1]][1], 
                                                          USE.NAMES = F))
    query.fields <- lapply(1:length(query.fields), function(i, 
                                                            qf = query.fields, qc = query.cont, qn = query.names) {
      x <- qf[[i]]
      x[which(is.na(x) == T)] <- sapply(qc[[i]][which(is.na(x) == 
                                                        T)], function(y) strsplit(strsplit(y, "href=\"")[[1]][2], 
                                                                                  "\"/>")[[1]][1], USE.NAMES = F)
      names(x) <- qn[[i]]
      return(x)
    })
    return.names <- unique(unlist(query.names))
    return.df <- as.data.frame(stats::setNames(replicate(length(return.names), 
                                                         numeric(0), simplify = F), return.names))
    return.df <- do.call(rbind.data.frame, lapply(query.fields, 
                                                  function(x, rn = return.names, rdf = return.df) {
                                                    rdf[1, match(names(x), rn)] <- sapply(x, as.character)
                                                    return(rdf)
                                                  }))
  }
  if (is.TRUE(give.return)) 
    return(return.df)
}

File: Porting/Porting API/Done/GetSpatialDataDev.jl
module GetSpatialDataDev

#Global variables
#= imports
  library(httr)
  library(xml2)
  source("getSpatialData_dev_internal.R")
=#

using DataFrames
using Dates

include("getSpatialDataDevInternal.jl")

export cophub_api, getSentinel_query2, login_CopHub, set_archive #Functions


function cophub_api(x::AbstractString, p::AbstractString, user, pw) 
  if x == "auto"
    x = ( p == "Sentinel-1" || p == "Sentinel-2" ) ? "operational" : "pre-ops"
  elseif x == "operational"
    x = options[:gSD_api][:dhus]
  elseif x == "pre-ops"
    x = options[:gSD_api][:s3]
    user = "s3guest"
    pw = "s3guest"
  end
  return [ user, pw, x ]
end




function set_archive(dir_archive) 
  if !isa(dir_archive, AbstractString)
    out( "Argument 'dir_archive' needs to be of type 'String'.", type = 3 )
  end
  if !isdir(dir_archive)
    out("The defined directory does not exist.", type = 3) 
  end
  options[:gSD_archive] = dir_archive
  options[:gSD_archive_set] = true
end




function login_CopHub(username; password = nothing) 
  if isnothing(password)
    password = getPass()
  end
  char_args = Dict( :username => username, :password => password )
  for (key, val) in char_args
    if !isa(val, AbstractString)
      out( "Argument '$key' needs to be of type 'String'.", type = 3 )
    end
  end
  options[:gSD_cophub_user] = username
  options[:gSD_cophub_pass] = password
  options[:gSD_cophub_set] = true
end




function getSentinel_query2(time_range, platform; aoi=nothing, username=nothing, password=nothing, hub="auto", verbose=true, ingestiondate=false, extra=nothing)
  if options[:gSD_cophub_set]
    if isnothing(username) 
      username = options[:gSD_cophub_user]
    end
    if isnothing(password) 
      password = options[:gSD_cophub_pass]
    end
  end

  if !isa(username, AbstractString) 
    println("Argument 'username' needs to be of type 'String'. You can use 'login_CopHub()' to define your login credentials globally.")
  end

  if isnothing(password)
    password = getPass()
  end

    if verbose <: Bool
      options[:verbose] = verbose
    #if isnothing(aoi)
    #  if options[:gSD_aoi_set]
    #    aoi = options[:gSD_aoi]
    #  else
    #    out("Argument 'aoi' is undefined and no session AOI could be obtained. Define aoi or use set_aoi() to define a session AOI.", type = 3)
    #  end
    #end
    #aoi = make_aoi(aoi, type = "matrix")
  end


  char_args = Dict( :time_range => time_range, :platform => platform )
  for (key, val) in char_args
    if !isa(val, AbstractString)
      out( "Argument '$key' needs to be of type 'character'.", type = 3)
    end
  end
  if length(time_range) != 2
    out( "Argument 'time_range' must contain two elements (start and stop time).", type = 3 )
  end


  function cop_url( ext_xy, urlRoot, platform, timeRange, rowStart )
    if !ingestiondate
      qs = Dict( :urlRoot => join([urlRoot, "/"]),
                 :search => ( "search?start=", "&rows=100&q=(" ), 
                 :and => "%20AND%20", 
                 :aoi_poly => ( "footprint:%22Intersects(POLYGON((", ")))%22" ), 
                 :platformName => "platformname:", 
                 :time => Dict( "[" => "beginposition:%5b", "to" => "%20TO%20", "]" => "%5d" ) )
    else
      qs = Dict( :urlRoot => join([urlRoot, "/"]),
                 :search => ( "search?start=", "&rows=100&q=(" ), 
                 :and => "%20AND%20", 
                 :aoi_poly => ( "footprint:%22Intersects(POLYGON((", ")))%22" ), 
                 :platformName => "platformname:", 
                 :time => Dict( "[" => "ingestiondate:%5b", "to" => "%20TO%20", "]" => "%5d" ) )
    end
  

    #aoi_str = join( join.( ext_xy, "%20" ), "," )

    timeRange = map( x -> tt = format( DateTime( timeRange ), "%Y-%m-%dT%H:%M:%S.000Z" ), timeRange )
    if isnothing(extra)
      urlstring = join([ qs[:urlRoot], (qs[:search])[1], rowStart,
                         (qs[:search])[2], qs[:platformName], platform,
                         qs[:and], (qs[:time])["["], timeRange[1],
                         (qs[:time])["to"], timeRange[2], (qs[:time])["]"], ")" ])
    else
      urlstring = join([ qs[:urlRoot], (qs[:search])[1], rowStart,
                         (qs[:search])[2], qs[:platformName], platform,
                         qs[:and], extra, qs[:and], (qs[:time])["["],
                         timeRange[1], (qs[:time])["to"], timeRange[2],
                         (qs[:time])["]"], ")" ])  
    end
    println(urlstring)
    println( now() )
    return urlstring
  end

  cred = cophub_api(hub, platform, username, password)
  row_start = -100
  re_query = true
  give_return = true
  query_list = []
  while re_query
    row_start += 100
    query = gSD_get( cop_url( aoi, cred[3], platform, time_range, row_start ), cred[1], cred[2] )
    query_xml = xml_contents(contents(query))
    append!( query_list, map( x -> xml_contents(x), filter(str -> contains("entry", str), query_xml) ) )


    if length(query_list) == 0 && row_start == 0
      out( "No results could be obtained for this request.", msg = true )
      re_query = false
      give_return = false
    end
    if length(query_list) != row_start + 100
      re_query = false
    end
  end






  if give_return
    field_tag = [ "title", "link href", "link rel=\"alternative\"", "link rel=\"icon\"", "summary", "name" ]
    field_names = [ "title", "url", "url.alt", "url.icon", "summary" ]

    query_cont = map( (x, field = field_tag) -> map( (f, y = x) -> filter( line ->  occursin(f, line), y ), field ), query_list )
    query_names = map( (x, field_n = field_names) -> append!( field_n, map( y -> split( y, '\"', keepempty = false )[2], x[length(field_n)+1 : end] ) ), query_cont )
    query_fields = map( x -> map( y -> split( split(y, '<', keepempty = false)[2], '>', keepempty = false )[1], x ), query_cont )
    
    function fun( i, qf, qc, qn )
        x = qf[i]
        x[ ismissing.(x) ] = map( y -> split( split(y, "href=\"")[2], "\">" )[1], qc[i][ ismissing.(x) .== true ] )
        #names(x) <- qn[[i]]
        return x
    end
    query_fields = map( (i, qf = query_fields, qc = query_cont, qn = query_names ) -> fun( i, qf, qc, qn ), 1:length(query_fields) )

    return_names = unique(query_names)
    return_df = rename!( convert( DataFrame, return_names  ), zeros( Int64,  length(return_names) ) )
    return_df = combine( DataFrame( map( (x, rn = return_names, rdf = return_df) -> rdf[1, findall( in(names(x)), rn )] = map( y -> string(y), x ) ) ) )
  end

  if give_return
    return return_df
  end
end

end # module